pkg/gca/endpoints/endpoints.go

```diff
@@ -0,0 +1,266 @@
+	// is canceled or any of the endpoints fails to run.
+	ListenAndServe(ctx context.Context) error
+}
+
+// Config represents the configuration of the Galadriel ServerCA Endpoints
+type Config struct {
+	// TPCAddr is the address to bind the TCP listener to.
+	TCPAddress *net.TCPAddr
+
+	// LocalAddress is the local address to bind the listener to.
+	LocalAddress net.Addr
+
+	Logger logrus.FieldLogger
+
+	// ServerCA is used for signing X.509 certificates and JWTs
+	ServerCA ca.ServerCA
+
+	JWTTokenTTL time.Duration
+	X509CertTTL time.Duration
+
+	Clock clock.Clock
+}
+
+type Endpoints struct {
+	serverCA   ca.ServerCA
+	tcpAddress *net.TCPAddr
+	localAddr  net.Addr
+	logger     logrus.FieldLogger
+	clock      clock.Clock
+	config     *Config
+
+	certsStore *tlsCertSource
+
+	hooks struct {
+		// test hook used to indicate that is listening on TCP
+		tcpListening chan struct{}
+	}
+}
+
+func New(c *Config) (*Endpoints, error) {
+	if err := util.PrepareLocalAddr(c.LocalAddress); err != nil {
+		return nil, err
+	}
+
+	return &Endpoints{
+		serverCA:   c.ServerCA,
+		tcpAddress: c.TCPAddress,
+		localAddr:  c.LocalAddress,
+		logger:     c.Logger,
+		clock:      c.Clock,
+		config:     c,
+	}, nil
+}
+
+func (e *Endpoints) ListenAndServe(ctx context.Context) error {
+	err := util.RunTasks(ctx,
+		e.runTCPServer,
+		e.runUDSServer,
+	)
+	if errors.Is(err, context.Canceled) {
+		err = nil
+	}
+
+	return err
+}
+
+type tlsCertSource struct {
+	mu   sync.RWMutex
+	cert *tls.Certificate
+}
+
+func (t *tlsCertSource) setCert(cert *tls.Certificate) {
+	t.mu.Lock()
+	defer t.mu.Unlock()
+	t.cert = cert
+}
+
+func (t *tlsCertSource) getCert() *tls.Certificate {
+	t.mu.Lock()
+	defer t.mu.Unlock()
+	return t.cert
+}
+
+func (e *Endpoints) runTCPServer(ctx context.Context) error {
+	var err error
+
+	cert, err := e.getTLSCertificate(ctx)
+	if err != nil {
+		return err
+	}
+
+	e.certsStore = &tlsCertSource{cert: cert}
+
+	tlsConfig := &tls.Config{
+		GetCertificate: func(info *tls.ClientHelloInfo) (*tls.Certificate, error) {
+			return e.certsStore.getCert(), nil
+		},
+	}
+
+	// TCP API handlers
+	mux := http.NewServeMux()
+
+	server := &http.Server{
+		Addr:      e.tcpAddress.String(),
+		Handler:   mux,
+		TLSConfig: tlsConfig,
+	}
+
+	e.logger.Infof("Starting secure GCA TCP endpoint listening on %s", e.tcpAddress.String())
+	errChan := make(chan error)
+	go func() {
+		e.triggerListeningHook()
+		// certificate and key are embedded in the TLS config
+		errChan <- server.ListenAndServeTLS("", "")
+	}()
+
+	go e.tlsCertificateRotator(ctx, errChan)
+
+	select {
+	case err = <-errChan:
+		e.logger.WithError(err).Error("GCA TCP endpoint stopped prematurely")
+		return err
+	case <-ctx.Done():
+		e.logger.Info("Stopping GCA TCP endpoint")
+		err = server.Close()
+		if err != nil {
+			e.logger.WithError(err).Error("Error closing GCA TCP endpoint")
+		}
+		<-errChan
+		e.logger.Info("GCA TCP endpoint stopped")
+		return nil
+	}
```

Comment:
Should we close the error channel? In the case of a yes. The UDS also needs that.
```suggestion
	select {
	case err = <-errChan:
		e.logger.WithError(err).Error("GCA TCP endpoint stopped prematurely")
		close(errChan)
		return err
	case <-ctx.Done():
		e.logger.Info("Stopping GCA TCP endpoint")
		err = server.Close()
		if err != nil {
			e.logger.WithError(err).Error("Error closing GCA TCP endpoint")
		}
		<-errChan
		e.logger.Info("GCA TCP endpoint stopped")
		return nil
	}
```

Reply:
I don't think we need this, as channels in Go are not resources like files or sockets that need to be explicitly closed to free up system resources. When a channel is no longer referenced by any goroutine, it becomes eligible for garbage collection by the Go runtime, and the memory used by the channel is automatically reclaimed.